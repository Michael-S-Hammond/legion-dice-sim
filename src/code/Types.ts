export enum DieColor {
    Red = "red",
    Black = "black",
    White = "white",
}

export enum AttackDieResult {
    Critical = 1,
    Hit = 2,
    Surge = 3,
    Miss = 4,
}

export type AttackRoll = {
  color: DieColor,
  result: AttackDieResult,
}

export enum AttackSurgeConversion {
  Blank = 1,
  Hit = 2,
  Critical = 3,
}

export enum Cover {
  None = "none",
  Light = "light",
  Heavy = "heavy",
}

export enum DefenseDieResult {
  Block = 1,
  Surge = 2,
  Blank = 3,
}

export type AbilityX = {
  active: boolean,
  value: number,
}

export type OffenseTokens = {
  aim: number,
  surge: number,
}

export type OffenseInput = {
  redDice: number,
  blackDice: number,
  whiteDice: number,
  surge: AttackSurgeConversion,
  tokens: OffenseTokens,
  blast: boolean,
  criticalX: AbilityX,
  duelist: boolean,
  highVelocity: boolean,
  impactX: AbilityX,
  ionX: AbilityX,
  jediHunter: boolean,
  lethalX: AbilityX,
  makashiMastery: boolean,
  pierceX: AbilityX,
  preciseX: AbilityX,
  ramX: AbilityX,
  sharpshooterX: AbilityX,
}

export type DefenseTokens = {
  dodge: number,
  observation: number,
  shield: number,
  suppression: number
  surge: number,
}

export type DefenseInput = {
  dieColor: DieColor,
  surge: boolean,
  tokens: DefenseTokens,
  cover: Cover,
  armor: boolean,
  armorX: AbilityX,
  block: boolean,
  dangerSenseX: AbilityX,
  deflect: boolean,
  djemSoMastery: boolean,
  duelist: boolean,
  hasForceUpgrades: boolean,
  immuneBlast: boolean,
  immuneMeleePierce: boolean,
  immunePierce: boolean,
  impervious: boolean,
  lowProfile: boolean,
  outmaneuver: boolean,
  soresuMastery: boolean,
  uncannyLuckX: AbilityX,
}

export type CombatInput = {
  guardian: AbilityX,
  meleeAttack: boolean,
};

export type AttackInput = {
  offense: OffenseInput,
  defense: DefenseInput,
  combat: CombatInput,
}

export type AttackResult = {
  criticals: number,
  hits: number,
  surges: number,
  misses: number,
}

export type DefenseResult = {
  forcedSaves: number,
  blocks: number,
  surges: number,
  blanks: number,
  wounds: number,
}

export type AttackOutput = {
  attack: AttackResult,
  defense: DefenseResult,
}

export type Stats = {
  mean: number,
  median: number,
  stddev: number,
}

export type AttackSummary = {
  attackCount: number,
  critical: Array<number>,
  hit: Array<number>,
  attackSurge: Array<number>,
  forcedSaves: Array<number>,
  forcedSaveStats: Stats,
  blocks: Array<number>,
  defenseSurge: Array<number>,
  wounds: Array<number>,
  woundStats: Stats
}

export type CombinedAttackOutput = {
  firstAttack: AttackOutput,
  summary: AttackSummary,
}

export enum ResultOutput {
  None = 1,
  Single = 2,
  Graph = 3,
}

export function createDefaultAttackInput() : AttackInput {
  return {
      offense: {
          redDice: 0,
          blackDice: 0,
          whiteDice: 0,
          surge: AttackSurgeConversion.Blank,
          tokens: {
            aim: 0,
            surge: 0,
          },
          blast: false,
          criticalX: { active: false, value: 1 },
          duelist: false,
          highVelocity: false,
          impactX: { active: false, value: 1 },
          ionX: { active: false, value: 1 },
          jediHunter: false,
          lethalX: { active: false, value: 1 },
          makashiMastery: false,
          pierceX: { active: false, value: 1 },
          preciseX: { active: false, value: 1 },
          ramX: { active: false, value: 1 },
          sharpshooterX: { active: false, value: 1 },
      },
      defense: {
          dieColor: DieColor.White,
          surge: true,
          tokens: {
            dodge: 0,
            observation: 0,
            shield: 0,
            suppression: 0,
            surge: 0,
          },
          cover: Cover.None,
          armor: false,
          armorX: { active: false, value: 1 },
          block: false,
          dangerSenseX: { active: false, value: 1 },
          deflect: false,
          djemSoMastery: false,
          duelist: false,
          hasForceUpgrades: false,
          immuneBlast: false,
          immuneMeleePierce: false,
          immunePierce: false,
          impervious: false,
          lowProfile: false,
          outmaneuver: false,
          soresuMastery: false,
          uncannyLuckX: { active: false, value: 1 }
      },
      combat: {
        guardian: { active: false, value: 1 },
        meleeAttack: false,
      },
  }
}

export function cloneAttackInput(original: AttackInput) : AttackInput {
  return {
      offense: {
        redDice: original.offense.redDice,
        blackDice: original.offense.blackDice,
        whiteDice: original.offense.whiteDice,
        surge: original.offense.surge,
        tokens: {
          aim: original.offense.tokens.aim,
          surge: original.offense.tokens.surge,
        },
        blast: original.offense.blast,
        criticalX: {
          active: original.offense.criticalX.active,
          value: original.offense.criticalX.value
        },
        duelist: original.offense.duelist,
        highVelocity: original.offense.highVelocity,
        impactX: {
          active: original.offense.impactX.active,
          value: original.offense.impactX.value
        },
        ionX: {
          active: original.offense.ionX.active,
          value: original.offense.ionX.value
        },
        jediHunter: original.offense.jediHunter,
        lethalX: {
          active: original.offense.lethalX.active,
          value: original.offense.lethalX.value
        },
        makashiMastery: original.offense.makashiMastery,
        pierceX: {
          active: original.offense.pierceX.active,
          value: original.offense.pierceX.value
        },
        preciseX: {
          active: original.offense.preciseX.active,
          value: original.offense.preciseX.value
        },
        ramX: {
          active: original.offense.ramX.active,
          value: original.offense.ramX.value,
        },
        sharpshooterX: {
          active: original.offense.sharpshooterX.active,
          value: original.offense.sharpshooterX.value
        },
    },
    defense: {
        dieColor: original.defense.dieColor,
        surge: original.defense.surge,
        tokens: {
          dodge: original.defense.tokens.dodge,
          observation: original.defense.tokens.observation,
          shield: original.defense.tokens.shield,
          suppression: original.defense.tokens.suppression,
          surge: original.defense.tokens.surge,
        },
        cover: original.defense.cover,
        armor: original.defense.armor,
        armorX: {
          active: original.defense.armorX.active,
          value: original.defense.armorX.value
        },
        block: original.defense.block,
        dangerSenseX: {
          active: original.defense.dangerSenseX.active,
          value: original.defense.dangerSenseX.value
        },
        deflect: original.defense.deflect,
        djemSoMastery: original.defense.djemSoMastery,
        duelist: original.defense.duelist,
        hasForceUpgrades: original.defense.hasForceUpgrades,
        immuneBlast: original.defense.immuneBlast,
        immuneMeleePierce: original.defense.immuneMeleePierce,
        immunePierce: original.defense.immunePierce,
        impervious: original.defense.impervious,
        lowProfile: original.defense.lowProfile,
        outmaneuver: original.defense.outmaneuver,
        soresuMastery: original.defense.soresuMastery,
        uncannyLuckX: {
          active: original.defense.uncannyLuckX.active,
          value: original.defense.uncannyLuckX.value
        },
    },
    combat: {
      guardian: {
        active: original.combat.guardian.active,
        value: original.combat.guardian.value
      },
      meleeAttack: original.combat.meleeAttack,
    },
  };
}
