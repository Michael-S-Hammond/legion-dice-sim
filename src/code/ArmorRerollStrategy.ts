import * as T from "./Types";
import * as RS from "./RerollStrategy";
import * as RSH from "./RerollStrategyHelpers";

import DiceModificationMatrix, * as DMM from "./DiceModificationMatrix";
import * as EC from './EffectiveCover';

class ArmorRerollStrategy implements RS.RerollStrategy {
    pickRerollDice(input: T.AttackInput,
                   rolls: T.AttackRoll[],
                   rerollDiceCount: number,
                   reason: RS.RerollReason,
                   remaining: RS.RemainingRerolls) : number[] | undefined {
        const dmm = new DiceModificationMatrix(rolls);
        const effectiveCover = EC.getEffectiveCover(input);
        const coverModifier = EC.getCoverModification(effectiveCover);
        let rerollMiss = 0;

        // convert surges
        if(input.offense.surge === T.AttackSurgeConversion.Critical) {
            dmm.tryConvertResultCount(T.AttackDieResult.Surge, T.AttackDieResult.Critical, rolls.length);
        }
        if(input.offense.criticalX.active) {
            dmm.tryConvertResultCount(T.AttackDieResult.Surge, T.AttackDieResult.Critical, input.offense.criticalX.value);
        }
        if(input.offense.surge === T.AttackSurgeConversion.Hit) {
            dmm.tryConvertResultCount(T.AttackDieResult.Surge, T.AttackDieResult.Hit, rolls.length);
        }
        dmm.tryConvertResultCount(T.AttackDieResult.Surge, T.AttackDieResult.Miss, rolls.length);

        // ram
        if(input.offense.ramX.active && input.combat.meleeAttack) {
            let ramRemaining = input.offense.ramX.value;
            ramRemaining -= dmm.tryConvertResultCount(T.AttackDieResult.Miss, T.AttackDieResult.Critical, ramRemaining);
            dmm.tryConvertResultCount(T.AttackDieResult.Hit, T.AttackDieResult.Critical, ramRemaining);
        }

        // any hits to be rerolled should be converted to misses in this block
        // therefore, hits should always be 0 when this completes

        const totalHits = dmm.getResultCount(T.AttackDieResult.Hit);
        let totalCrits = dmm.getResultCount(T.AttackDieResult.Critical);
        const possibleHits = Math.min(rolls.length - totalCrits, totalHits + RSH.getMaximumRerolledDice(input, remaining));
        let hitsToKeep = 0;
        if(input.offense.impactX.active && (possibleHits > coverModifier)) {
            hitsToKeep = Math.min(totalHits, coverModifier + input.offense.impactX.value);
        }

        // remove hits for cover and armor, less impact
        dmm.tryConvertResultAllExcept(T.AttackDieResult.Hit, T.AttackDieResult.Miss, hitsToKeep);

        // calculate misses that can be rerolled
        rerollMiss = Math.min(rerollDiceCount, dmm.getResultCount(T.AttackDieResult.Miss));

        if(reason === RS.RerollReason.AimToken) {
            // convert hits to crit for impact
            const impactConvert = Math.max(0, dmm.getResultCount(T.AttackDieResult.Hit) - coverModifier);
            dmm.tryConvertResultCount(T.AttackDieResult.Hit, T.AttackDieResult.Critical, impactConvert);

            // lethal
            totalCrits = dmm.getResultCount(T.AttackDieResult.Critical);
            const effectiveLethal = RSH.getEffectiveLethal(input, 0, totalCrits, 0);
            if(effectiveLethal > 0 && remaining.aimTokens <= effectiveLethal) {
                rerollMiss = 0;
            }
        }

        // return the result
        const indexes = dmm.getRerollIndexes(0, rerollMiss);
        return indexes.length == 0 ? undefined : indexes;
    }

    shouldHandle(input: T.AttackInput, reason: RS.RerollReason) : boolean {
        return input.defense.armor;
    }
}

export default ArmorRerollStrategy;
