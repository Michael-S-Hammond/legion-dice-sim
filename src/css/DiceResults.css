.single-result {
  display: flex; }

.crit-result-img {
  content: url("../img/CriticalHit.png");
  height: 30px;
  width: 30px; }

.hit-result-img {
  content: url("../img/Hit.png");
  height: 30px;
  width: 30px; }

.attack-surge-result-img {
  content: url("../img/AttackSurge.png");
  height: 30px;
  width: 30px; }

.block-result-img {
  content: url("../img/Block.png");
  height: 30px;
  width: 30px; }

.defense-surge-result-img {
  content: url("../img/DefenseSurge.png");
  height: 30px;
  width: 30px; }

.die-result {
  font-weight: bold;
  font-size: larger; }

.stats-label {
  font-size: large; }

.stats-header {
  text-align: center; }
