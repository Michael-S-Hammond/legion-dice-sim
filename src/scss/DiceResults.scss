$attack-img-size: 30px;

.single-result {
    display: flex;
}

@mixin result-img {
    height: $attack-img-size;
    width: $attack-img-size;
}
.crit-result-img {
    content: url('../img/CriticalHit.png');
    @include result-img();
}

.hit-result-img {
    content: url('../img/Hit.png');
    @include result-img();
}

.attack-surge-result-img {
    content: url('../img/AttackSurge.png');
    @include result-img();
}

.block-result-img {
    content: url('../img/Block.png');
    @include result-img();
}

.defense-surge-result-img {
    content: url('../img/DefenseSurge.png');
    @include result-img();
}

.die-result {
    font-weight: bold;
    font-size: larger;
}

.stats-label {
    font-size: large;
}

.stats-header {
    text-align: center;
}
