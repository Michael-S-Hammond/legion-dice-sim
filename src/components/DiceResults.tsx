import '../css/DiceResults.css';

import React from 'react';
import * as mjs from 'mathjs';
import * as Highcharts from 'highcharts';
import HighchartsReact from 'highcharts-react-official';
import ExpMod from 'highcharts/modules/exporting';
import OffExpMod from 'highcharts/modules/offline-exporting';

import * as T from '../code/Types'

ExpMod(Highcharts);
OffExpMod(Highcharts);

type DiceResultsProps = {
    results: T.CombinedAttackOutput,
    visibility: T.ResultOutput,
    showExpectedRange: boolean,
};

type DiceResultsState = {
    chartOptions: Highcharts.Options,
};

class DiceResults extends React.Component<DiceResultsProps, DiceResultsState> {
    private rootElementRef: React.RefObject<HTMLDivElement>;

    private critColor = 'rgba(192, 190, 195, 0.8)';
    private hitColor = 'rgba(50, 140, 193, 0.8)';
    private attackSurgeColor = 'rgba(11, 60, 93, 0.8)';
    private forcedSavesColor = 'rgba(14, 19, 21, 0.8)';
    private forcedSavesFillColor = 'rgba(14, 19, 21, 0.4)';
    private blocksColor = 'rgba(216, 108, 35, 0.8)';
    private defenseSurgeColor = 'rgba(255, 225, 0, 0.8)';
    private woundsColor = 'rgba(192, 45, 45, 0.8)';
    private woundsFillColor = 'rgba(192, 45, 45, 0.4)';
    private transparentColor = 'rgba(255, 255, 255, 0)';

    private refreshChart = true;
    private scrollUIIntoView = false;

    constructor(props: DiceResultsProps) {
        super(props);
        this.state = {
            chartOptions: this.createChartOptions(),
        };
        this.rootElementRef = React.createRef();
    }

    private getFirstCritCount(): string {
        return String(this.props.results.firstAttack.attack.criticals);
    }

    private getFirstHitCount(): string {
        return String(this.props.results.firstAttack.attack.hits);
    }

    private getFirstAttackSurgeCount(): string {
        return String(this.props.results.firstAttack.attack.surges);
    }

    private getFirstBlockCount(): string {
        return String(this.props.results.firstAttack.defense.blocks);
    }

    private getFirstDefenseSurgeCount(): string {
        return String(this.props.results.firstAttack.defense.surges);
    }

    private getFirstWoundCount(): string {
        return String(this.props.results.firstAttack.defense.wounds);
    }

    private createChartOptions(): Highcharts.Options {
        const chartData = [
            this.props.results.summary.critical,
            this.props.results.summary.hit,
            this.props.results.summary.attackSurge,
            this.props.results.summary.forcedSaves,
            this.props.results.summary.blocks,
            this.props.results.summary.defenseSurge,
            this.props.results.summary.wounds,
        ];
        const forcedSaveStats = this.props.results.summary.forcedSaveStats;
        const woundStats = this.props.results.summary.woundStats;
        return {
            chart: {
                type: 'spline',
            },
            title: {
                text: undefined,
            },
            credits: {
                enabled: false,
            },
            xAxis: {
                categories: this.getXAxisCategories(),
            },
            yAxis: {
                title: undefined,
            },
            series: [{
                type: 'spline',
                name: 'Crits',
                color: this.critColor,
                visible: false,
                data: chartData[0],
            }, {
                type: 'spline',
                name: 'Hits',
                color: this.hitColor,
                visible: false,
                data: chartData[1],
            }, {
                type: 'spline',
                name: 'Attack Surges',
                color: this.attackSurgeColor,
                visible: false,
                data: chartData[2]
            }, {
                type: 'areaspline',
                name: 'Required Saves',
                color: this.forcedSavesColor,
                fillColor: this.forcedSavesFillColor,
                zoneAxis: 'x',
                zones: this.getZones(forcedSaveStats.mean, forcedSaveStats.stddev),
                data: chartData[3],
            }, {
                type: 'spline',
                name: 'Blocks',
                color: this.blocksColor,
                visible: false,
                data: chartData[4],
            }, {
                type: 'spline',
                name: 'Defense Surges',
                color: this.defenseSurgeColor,
                visible: false,
                data: chartData[5],
            }, {
                type: 'areaspline',
                name: 'Wounds',
                color: this.woundsColor,
                fillColor: this.woundsFillColor,
                zoneAxis: 'x',
                zones: this.getZones(woundStats.mean, woundStats.stddev),
                data: chartData[6],
            }],
            plotOptions: {
                areaspline: {
                    marker: {
                        symbol: 'circle',
                    }
                },
                spline: {
                    marker: {
                        symbol: 'circle',
                    }
                },
            },
        };
    }

    componentDidUpdate() : void{
        if (this.props.visibility === T.ResultOutput.Graph) {
            this.updateChart();
        }
    }

    private getXAxisCategories(): string[] {
        const categories: string[] = [];
        for (let i = 0; i < this.props.results.summary.critical.length; i++) {
            categories.push(String(i));
        }
        return categories;
    }

    private getZones(mean: number, stddev: number) {
        if (this.props.showExpectedRange) {
            return [{
                fillColor: this.transparentColor,
                value: mean - stddev,
            }, {
                value: mean + stddev,
            }, {
                fillColor: this.transparentColor,
            }]
        } else {
            return [{
                fillColor: this.transparentColor,
                value: this.props.results.summary.critical.length,
            }];
        }
    }

    private updateChart() {
        if (this.refreshChart === false) {
            return;
        }

        this.setState({
            chartOptions: this.createChartOptions(),
        });
        this.refreshChart = false;

        this.performScrollIntoView();
    }

    public forceChartUpdate() : void {
        this.refreshChart = true;
    }

    public scrollIntoView() : void {
        this.scrollUIIntoView = true;
    }

    private performScrollIntoView() {
        if (this.scrollUIIntoView) {
            this.rootElementRef.current?.scrollIntoView({ behavior: "smooth" });
        }
        this.scrollUIIntoView = false;
    }

    render() : JSX.Element {
        return (
            <div ref={this.rootElementRef}>
                <div className={`single-result justify-content-center my-3 ${this.props.visibility === T.ResultOutput.Single ? 'collapse.show' : 'collapse'}`}>
                    <img className='crit-result-img mx-1'></img>
                    <span className="die-result">{this.getFirstCritCount()}</span>
                    <img className='hit-result-img mx-1'></img>
                    <span className="die-result">{this.getFirstHitCount()}</span>
                    <img className='attack-surge-result-img mx-1'></img>
                    <span className="die-result">{this.getFirstAttackSurgeCount()}</span>
                </div>
                <div className={`single-result justify-content-center my-3 ${this.props.visibility === T.ResultOutput.Single ? 'collapse.show' : 'collapse'}`}>
                    <img className='block-result-img mx-1'></img>
                    <span className="die-result">{this.getFirstBlockCount()}</span>
                    <img className='defense-surge-result-img mx-1'></img>
                    <span className="die-result">{this.getFirstDefenseSurgeCount()}</span>
                </div>
                <div className={`single-result justify-content-center my-3 ${this.props.visibility === T.ResultOutput.Single ? 'collapse.show' : 'collapse'}`}>
                    <span className="die-result">Wounds: {this.getFirstWoundCount()}</span>
                </div>
                <div id="chartContainer" className={`my-3 mx-auto ${this.props.visibility === T.ResultOutput.Graph ? 'collapse.show' : 'collapse'}`}>
                    <HighchartsReact
                        highcharts={Highcharts}
                        options={this.state.chartOptions}
                    ></HighchartsReact>
                </div>
                <div className={`row ${this.props.visibility === T.ResultOutput.Graph ? 'collapse.show' : 'collapse'}`}>
                    <div className="col-md-3 offset-md-3 my-3">
                        <h3 className="stats-header">Required Blocks</h3>
                        <div>
                            <span className="stats-label">Median: </span>
                            <span>{this.props.results.summary.forcedSaveStats.median}</span>
                        </div>
                        <div>
                            <span className="stats-label">Mean: </span>
                            <span>{mjs.round(this.props.results.summary.forcedSaveStats.mean, 3)}</span>
                        </div>
                        <div>
                            <span className="stats-label">Standard Deviation: </span>
                            <span>{mjs.round(this.props.results.summary.forcedSaveStats.stddev, 3)}</span>
                        </div>
                    </div>
                    <div className="col-md-3 my-3">
                        <h3 className="stats-header">Wounds</h3>
                        <div>
                            <span className="stats-label">Median: </span>
                            <span>{this.props.results.summary.woundStats.median}</span>
                        </div>
                        <div>
                            <span className="stats-label">Mean: </span>
                            <span>{mjs.round(this.props.results.summary.woundStats.mean, 3)}</span>
                        </div>
                        <div>
                            <span className="stats-label">Standard Deviation: </span>
                            <span>{mjs.round(this.props.results.summary.woundStats.stddev, 3)}</span>
                        </div>
                    </div>
                </div>
            </div>
        );
    }
}

export default DiceResults;
