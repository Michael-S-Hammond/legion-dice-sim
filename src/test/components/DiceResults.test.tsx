import React from 'react';

jest.mock('highcharts');

import { shallow } from 'enzyme';
import { Json, shallowToJson } from 'enzyme-to-json';

import * as T from '../../code/Types';
import DiceResults from '../../components/DiceResults';

jest.mock('highcharts/modules/exporting');
jest.mock('highcharts/modules/offline-exporting');

const removeChartFromSnapshot = (json: Json) : Json => {
    if(json.props['id'] === 'chartContainer') {
        return {
            ...json,
            children: [{
                type: 'Highchart',
                children: [],
                props: {},
                $$typeof: json.children[0].$$typeof,
            }]
        };
    }
    return json;
};

function createEmptyOutput() : T.CombinedAttackOutput {
    return {
        firstAttack: {
            attack: {
                criticals: 0,
                hits: 0,
                surges: 0,
                misses: 0,
            },
            defense: {
                blocks: 0,
                surges: 0,
                blanks: 0,
                forcedSaves: 0,
                wounds: 0,
            },
        },
        summary: {
            critical: [],
            hit: [],
            attackSurge: [],
            attackCount: 0,
            blocks: [],
            defenseSurge: [],
            forcedSaves: [],
            wounds: [],
            forcedSaveStats: {
                mean: 0,
                median: 0,
                stddev: 0,
            },
            woundStats: {
                mean: 0,
                median: 0,
                stddev: 0,
            },
        },
    };
}

describe('DiceResults', () => {
    it('matches the snapshot when not visible', () => {
        const output : T.CombinedAttackOutput = createEmptyOutput();

        const diceResults = shallow(<DiceResults
            showExpectedRange={true}
            visibility={T.ResultOutput.None}
            results={output}
            ></DiceResults>);

        const snapshot = shallowToJson(diceResults, { map: removeChartFromSnapshot });
        expect(snapshot).toMatchSnapshot();
    });

    it('matches the snapshot when showing single result', () => {
        const output : T.CombinedAttackOutput = createEmptyOutput();

        const diceResults = shallow(<DiceResults
            showExpectedRange={true}
            visibility={T.ResultOutput.Single}
            results={output}
            ></DiceResults>);

        const snapshot = shallowToJson(diceResults, { map: removeChartFromSnapshot });
        expect(snapshot).toMatchSnapshot();
    });

    it('matches the snapshot when showing graph result', () => {
        const output : T.CombinedAttackOutput = createEmptyOutput();

        const diceResults = shallow(<DiceResults
            showExpectedRange={true}
            visibility={T.ResultOutput.Graph}
            results={output}
            ></DiceResults>);

        const snapshot = shallowToJson(diceResults, { map: removeChartFromSnapshot });
        expect(snapshot).toMatchSnapshot();
    });
});
