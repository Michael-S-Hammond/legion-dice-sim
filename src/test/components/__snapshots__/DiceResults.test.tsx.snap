// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiceResults matches the snapshot when not visible 1`] = `
<div>
  <div
    className="single-result justify-content-center my-3 collapse"
  >
    <img
      className="crit-result-img mx-1"
    />
    <span
      className="die-result"
    >
      0
    </span>
    <img
      className="hit-result-img mx-1"
    />
    <span
      className="die-result"
    >
      0
    </span>
    <img
      className="attack-surge-result-img mx-1"
    />
    <span
      className="die-result"
    >
      0
    </span>
  </div>
  <div
    className="single-result justify-content-center my-3 collapse"
  >
    <img
      className="block-result-img mx-1"
    />
    <span
      className="die-result"
    >
      0
    </span>
    <img
      className="defense-surge-result-img mx-1"
    />
    <span
      className="die-result"
    >
      0
    </span>
  </div>
  <div
    className="single-result justify-content-center my-3 collapse"
  >
    <span
      className="die-result"
    >
      Wounds: 
      0
    </span>
  </div>
  <div
    className="my-3 mx-auto collapse"
    id="chartContainer"
  >
    <Highchart />
  </div>
  <div
    className="row collapse"
  >
    <div
      className="col-md-3 offset-md-3 my-3"
    >
      <h3
        className="stats-header"
      >
        Required Blocks
      </h3>
      <div>
        <span
          className="stats-label"
        >
          Median: 
        </span>
        <span>
          0
        </span>
      </div>
      <div>
        <span
          className="stats-label"
        >
          Mean: 
        </span>
        <span>
          0
        </span>
      </div>
      <div>
        <span
          className="stats-label"
        >
          Standard Deviation: 
        </span>
        <span>
          0
        </span>
      </div>
    </div>
    <div
      className="col-md-3 my-3"
    >
      <h3
        className="stats-header"
      >
        Wounds
      </h3>
      <div>
        <span
          className="stats-label"
        >
          Median: 
        </span>
        <span>
          0
        </span>
      </div>
      <div>
        <span
          className="stats-label"
        >
          Mean: 
        </span>
        <span>
          0
        </span>
      </div>
      <div>
        <span
          className="stats-label"
        >
          Standard Deviation: 
        </span>
        <span>
          0
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`DiceResults matches the snapshot when showing graph result 1`] = `
<div>
  <div
    className="single-result justify-content-center my-3 collapse"
  >
    <img
      className="crit-result-img mx-1"
    />
    <span
      className="die-result"
    >
      0
    </span>
    <img
      className="hit-result-img mx-1"
    />
    <span
      className="die-result"
    >
      0
    </span>
    <img
      className="attack-surge-result-img mx-1"
    />
    <span
      className="die-result"
    >
      0
    </span>
  </div>
  <div
    className="single-result justify-content-center my-3 collapse"
  >
    <img
      className="block-result-img mx-1"
    />
    <span
      className="die-result"
    >
      0
    </span>
    <img
      className="defense-surge-result-img mx-1"
    />
    <span
      className="die-result"
    >
      0
    </span>
  </div>
  <div
    className="single-result justify-content-center my-3 collapse"
  >
    <span
      className="die-result"
    >
      Wounds: 
      0
    </span>
  </div>
  <div
    className="my-3 mx-auto collapse.show"
    id="chartContainer"
  >
    <Highchart />
  </div>
  <div
    className="row collapse.show"
  >
    <div
      className="col-md-3 offset-md-3 my-3"
    >
      <h3
        className="stats-header"
      >
        Required Blocks
      </h3>
      <div>
        <span
          className="stats-label"
        >
          Median: 
        </span>
        <span>
          0
        </span>
      </div>
      <div>
        <span
          className="stats-label"
        >
          Mean: 
        </span>
        <span>
          0
        </span>
      </div>
      <div>
        <span
          className="stats-label"
        >
          Standard Deviation: 
        </span>
        <span>
          0
        </span>
      </div>
    </div>
    <div
      className="col-md-3 my-3"
    >
      <h3
        className="stats-header"
      >
        Wounds
      </h3>
      <div>
        <span
          className="stats-label"
        >
          Median: 
        </span>
        <span>
          0
        </span>
      </div>
      <div>
        <span
          className="stats-label"
        >
          Mean: 
        </span>
        <span>
          0
        </span>
      </div>
      <div>
        <span
          className="stats-label"
        >
          Standard Deviation: 
        </span>
        <span>
          0
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`DiceResults matches the snapshot when showing single result 1`] = `
<div>
  <div
    className="single-result justify-content-center my-3 collapse.show"
  >
    <img
      className="crit-result-img mx-1"
    />
    <span
      className="die-result"
    >
      0
    </span>
    <img
      className="hit-result-img mx-1"
    />
    <span
      className="die-result"
    >
      0
    </span>
    <img
      className="attack-surge-result-img mx-1"
    />
    <span
      className="die-result"
    >
      0
    </span>
  </div>
  <div
    className="single-result justify-content-center my-3 collapse.show"
  >
    <img
      className="block-result-img mx-1"
    />
    <span
      className="die-result"
    >
      0
    </span>
    <img
      className="defense-surge-result-img mx-1"
    />
    <span
      className="die-result"
    >
      0
    </span>
  </div>
  <div
    className="single-result justify-content-center my-3 collapse.show"
  >
    <span
      className="die-result"
    >
      Wounds: 
      0
    </span>
  </div>
  <div
    className="my-3 mx-auto collapse"
    id="chartContainer"
  >
    <Highchart />
  </div>
  <div
    className="row collapse"
  >
    <div
      className="col-md-3 offset-md-3 my-3"
    >
      <h3
        className="stats-header"
      >
        Required Blocks
      </h3>
      <div>
        <span
          className="stats-label"
        >
          Median: 
        </span>
        <span>
          0
        </span>
      </div>
      <div>
        <span
          className="stats-label"
        >
          Mean: 
        </span>
        <span>
          0
        </span>
      </div>
      <div>
        <span
          className="stats-label"
        >
          Standard Deviation: 
        </span>
        <span>
          0
        </span>
      </div>
    </div>
    <div
      className="col-md-3 my-3"
    >
      <h3
        className="stats-header"
      >
        Wounds
      </h3>
      <div>
        <span
          className="stats-label"
        >
          Median: 
        </span>
        <span>
          0
        </span>
      </div>
      <div>
        <span
          className="stats-label"
        >
          Mean: 
        </span>
        <span>
          0
        </span>
      </div>
      <div>
        <span
          className="stats-label"
        >
          Standard Deviation: 
        </span>
        <span>
          0
        </span>
      </div>
    </div>
  </div>
</div>
`;
